pylint/checkers/typecheck.py

```diff
@@ -74,6 +74,7 @@
 import astroid
 import astroid.exceptions
 from astroid import bases, nodes
+from astroid.brain import brain_dataclasses
```

Comment:
I'm wondering if we should add this to a `utils` module in `astroid`. Porting the util to `pylint` doesn't make sense as we would need to duplicate the globals that are being used in the function, but importing from `brain` also feels weird.

@Pierre-Sassoulas Do you have an opinion?

Reply:
Importing the brain do feel weird, we're accessing astroid internals.

Reply:
Maybe this should be moved in pylint, like we did for ``is_typing_guard``, but I did not check if it's used in astroid itself.

Reply:
It's used in `astroid` and the function would be impacted by https://github.com/PyCQA/astroid/pull/1298 for example.

We need to decide where we want to store our "list of dataclass-like decorators". I think `astroid` makes the most sense (also because it is already there). However, I think a refactor into a `utils` module which is explicitly public API makes sense.

Reply:
Quoting my review in https://github.com/PyCQA/astroid/pull/1298#pullrequestreview-834665567

> There might be a problem with our design if we need to add the name of all dataclasses modules to this constant. I don't have a better solution as it's astroid and not pylint we can't add an option to let user add their own data-classes.

I don't know about the feasibility of moving this to pylint though.

Maybe we should keep it in astroid but provide an argument to be able to add user defined dataclasses (for example from a pylint option) ? @cdce8p what do you think ?

Reply:
I haven't had time to look at it in detail yet, but I'm not sure it's even necessary here. The goal of inference tips in astroid, as we have for `dataclasses` is to "hide" the magic and convert it to a format we normally expect.

If you need to check that it's not a decorator in a dataclass, probably the astroid brain would need to be improved. In this case I would think however, it's just not necessary. @jacobtylerwalls Could you add a test case for it?

Reply:
Hi, just checking I'm understanding -- you're asking for a quick demo that the line is needed? When I comment out the check I get these seven failures:

```
====================================================== short test summary info ======================================================
FAILED tests/test_functional.py::test_functional[unpacking_non_sequence_py37] - AssertionError: Wrong results for file "unpacking_...
FAILED tests/test_functional.py::test_functional[unspecified_encoding_py38] - AssertionError: Wrong results for file "unspecified_...
FAILED tests/test_functional.py::test_functional[no_member_dataclasses] - AssertionError: Wrong results for file "no_member_datacl...
FAILED tests/test_functional.py::test_functional[no_member_subclassed_dataclasses] - AssertionError: Wrong results for file "no_me...
FAILED tests/test_functional.py::test_functional[too_many_instance_attributes_py37] - AssertionError: Wrong results for file "too_...
FAILED tests/test_functional.py::test_functional[too_few_public_methods_37] - AssertionError: Wrong results for file "too_few_publ...
FAILED tests/test_functional.py::test_functional[dataclass_typecheck] - AssertionError: Wrong results for file "dataclass_typecheck":
============================================ 7 failed, 635 passed, 21 skipped in 33.58s =============================================
```

Which makes sense, because dataclasses define members with what looks like unused type annotations. For instance, several dozen here:

```
_______________________________________________ test_functional[dataclass_typecheck] ________________________________________________

self = <pylint.testutils.lint_module_test.LintModuleTest object at 0x7f9d0ba10f70>

    def runTest(self) -> None:
>       self._runTest()
E       AssertionError: Wrong results for file "dataclass_typecheck":
E       
E       Unexpected in testdata:
E         30: no-member
E         31: no-member
E         32: no-member
E         34: no-member
E         35: no-member
E         36: no-member
E         38: no-member
E         39: no-member
E         40: no-member
E         42: no-member
E         43: no-member
...
```